### PR TITLE
[tvmc] Introduce 'tune' subcommand (part 3/4)

### PR DESCRIPTION
**Pull request by leandron**
Introduce a sub-command to drive auto-tuning via command line, using AutoTVM, to produce the tuning records that can be used at compile time, or to further improve other tuning sessions. It follow good practices present on the auto-tuning tutorials.

In case you want to try it out (supposing you have your tvm built and working with this PR), you can have a look on the example below:
```
$ wget https://github.com/onnx/models/raw/master/vision/classification/resnet/model/resnet50-v2-7.onnx
$ python -m tvm.driver.tvmc --verbose tune --trials 1 --output log1.txt --target="llvm" resnet50-v2-7.onnx
(...messages...)
WARNING:root:Attribute spatial is ignored in relay.sym.batch_norm
[Task  1/23]  Current/Best:    0.00/   0.00 GFLOPS | Progress: (1/1) | 1.32 s Done.
[Task  2/23]  Current/Best:    0.00/   0.00 GFLOPS | Progress: (1/1) | 1.23 s Done.
[Task  3/23]  Current/Best:    0.00/   0.00 GFLOPS | Progress: (1/1) | 1.07 s Done.
...
```

cc @comaniac @masahi @FrozenGene for reviews if possible

### CODE REVIEW

File: python/tvm/driver/tvmc/autotuner.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    parser.set_defaults(func=drive_tune)
+    parser.add_argument(
+        "--early-stopping",
+        type=int,
+        help="minimum number of trials before early stopping",
+    )
+    parser.add_argument("--hostname", help="hostname or IP address of the host machine")
+    parser.add_argument(
+        "--min-repeat-ms",
+        default=1000,
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:50]: The current suggested numbers would be 0 for x86 and 1000 for other targets. We might need to reflect the best practice.

File: python/tvm/driver/tvmc/autotuner.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        required=True,
+        help="output file to store the tuning records for the tuning process",
+    )
+    parser.add_argument(
+        "--parallel",
+        default=4,
+        type=int,
+        help="the maximum number of parallel devices to use when tuning",
+    )
+    parser.add_argument("--port", default=9090, type=int, help="the port to connect to")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:79]: Would be better to put it along with the hostname.

File: python/tvm/driver/tvmc/autotuner.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        type=int,
+        default=1,
+        help="how many times to repeat each measurement",
+    )
+    parser.add_argument(
+        "--target",
+        help="compilation target as plain string, inline JSON or path to a JSON file",
+        required=True,
+    )
+    parser.add_argument("--timeout", default=10, help="time in seconds before timing out a config")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:91]: Build timeout, measurement timeout, or both?

File: python/tvm/driver/tvmc/autotuner.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    parser.add_argument("--tracker-key", help="the tracker key of the target device")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:92]: Put this along with hostname and port. We can even consider creating an argument group for them.

File: python/tvm/driver/tvmc/autotuner.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    Parameters
+    ----------
+    args: argparse.Namespace
+        Arguments from command line parser.
+
+    Returns
+    --------
+    int
+        Zero if successfully completed
+
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:134]: remove this line.

File: python/tvm/driver/tvmc/autotuner.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    Parameters
+    ----------
+    args: argparse.Namespace
+        Arguments from command line parser.
+
+    Returns
+    --------
+    int
+        Zero if successfully completed
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:133]: Explicitly say this is the error code.

File: python/tvm/driver/tvmc/autotuner.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    mod, params = frontends.load_model(args.FILE, args.model_format)
+
+    tasks = get_tuning_tasks(
+        mod=mod,
+        params=params,
+        target=target,
+        alter_layout=args.desired_layout,
+    )
+
+    if args.hostname:
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:147]: We may need to figure out a better way to let user indicate whether to use RPC runner. Maybe we can use a single argument like
```
--rpc-runner=0.0.0.0:9099-rasp4b
```
@FrozenGene do you have an idea?

File: python/tvm/driver/tvmc/autotuner.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        "early_stopping": args.early_stopping,
+        "measure_option": autotvm.measure_option(
+            builder=autotvm.LocalBuilder(build_func="default"), runner=runner
+        ),
+        "tuning_records": args.tuning_records,
+    }
+    logging.debug(" tuning options: %s", tuning_option)
+
+    tune_tasks(tasks, args.output, **tuning_option)
+    return 0
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:187]: Looks like 0 is the only possible return value, since other cases will directly raise exceptions. If so, it seems not necessary to have a return value.

File: python/tvm/driver/tvmc/autotuner.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    alter_layout : str, optional
+        The layout to convert the graph to. Note, the convert layout
+        pass doesn't currently guarantee the whole of the graph will
+        be converted to the chosen layout.
+
+    Returns
+    -------
+    tasks : list of autotvm.Tasks
+        list of tasks to be tuned
+
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:212]: remove this line.

File: python/tvm/driver/tvmc/autotuner.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    """
+    if not target_host:
+        target_host = target
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:215]: Is this correct? What if `target` is `cuda`?

File: python/tvm/driver/tvmc/autotuner.py [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+    if alter_layout:
+        mod = common.convert_graph_layout(mod, alter_layout)
+
+    tasks = autotvm.task.extract_from_program(
+        mod["main"],
+        target=target,
+        target_host=target_host,
+        params=params,
+        ops=(relay.op.get("nn.conv2d"),),
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:225]: Why only extracting conv2d tasks?

File: python/tvm/driver/tvmc/autotuner.py [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+        elif tuner == "random":
+            tuner_obj = RandomTuner(tsk)
+        elif tuner == "gridsearch":
+            tuner_obj = GridSearchTuner(tsk)
+        else:
+            raise TVMCException("invalid tuner: %s " % tuner)
+
+        # If transfer learning is being used, load the existing results
+        if tuning_records and os.path.exists(tuning_records):
+            logging.warning("loading tuning records from %s", tuning_records)
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:288]: `info` should be a proper logging level here.

File: python/tvm/driver/tvmc/autotuner.py [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+            start_time = time.time()
+            tuner_obj.load_history(autotvm.record.load_from_file(tuning_records))
+            logging.info("loaded history in %s", str(time.time() - start_time))
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:291]:
```suggestion
            logging.info("loaded history in %.2f sec(s)", time.time() - start_time)
```

File: python/tvm/driver/tvmc/common.py [hunk 14]
```diff
@@ -17,6 +17,11 @@
 """
 Common utility functions shared by TVMC modules.
 """
+import logging
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:20]: Use the standard logging, such as `logger = logging.getLogger("TVMC")`.

**Review by comaniac — CHANGES_REQUESTED** [review]
Also cc @FrozenGene @kevinthesun

File: python/tvm/driver/tvmc/autotuner.py (hunk 2 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:79]: I agree that merging a host name with optional port in a single parameter, e.g. `hostname[:port]` would be beneficial. However, I think that also adding the key to the same network-related parameter, would be a bit more complicated and obscure, in terms of syntax, making it harder for an end user to grasp.

Based on that, would you agree that for now we keep `hostname[:port]` in a single field, and the keep the key in a separate parameter?

File: python/tvm/driver/tvmc/autotuner.py (hunk 11 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:225]: I thought about that too. Do we have any documentation that can point to other tasks we might be interested? Interestingly on tutorials we only extract convolutions.

File: python/tvm/driver/tvmc/autotuner.py (hunk 4 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:92]: My considerations at https://github.com/apache/incubator-tvm/pull/6537#discussion_r493871858.

File: python/tvm/driver/tvmc/autotuner.py (hunk 2 above)
Comment (comaniac) [reply python/tvm/driver/tvmc/autotuner.py:79]: Fair enough. Then I'd suggest having a post-processing to check if `rpc-tracker` and `device-key` (need to think about the naming a bit more) are given or none of them. Similar to [this case](https://stackoverflow.com/questions/14350343/python-argparse-either-both-optional-arguments-or-else-neither-one).

File: python/tvm/driver/tvmc/autotuner.py (hunk 11 above)
Comment (comaniac) [reply python/tvm/driver/tvmc/autotuner.py:225]: Just don't specify this argument then it by default extracts all tasks.

File: python/tvm/driver/tvmc/autotuner.py (hunk 4 above)
Comment (FrozenGene) [reply python/tvm/driver/tvmc/autotuner.py:92]: How about name `--tracker-key` into `--rpc-device-key`?

File: python/tvm/driver/tvmc/autotuner.py (hunk 7 above)
Comment (FrozenGene) [reply python/tvm/driver/tvmc/autotuner.py:147]: I think we could check if `--rpc-device-key` (`--tracker-key` is in this pr, suggest naming it into `--rpc-device-key` so that we know it explicitly) is `None` or not. If it is not `None`, we are running RPC.

**Comment by FrozenGene** [comment]
Thanks @leandron , great job. Besides this, I think we could also add some features I think it is important too:
1. We could add `--num-threads` to control how many cores to tune. This is important for remote devices like ARM
2. We could add `--cpu-mode` to control BIG core or LITTLE core to tune. Imagine remote devices like have A72 + A35. This feature maybe should require us to expose one interface in our AutoTVM to use API `runtime.config_threadpool`.

File: python/tvm/driver/tvmc/autotuner.py (hunk 2 above)
Comment (comaniac) [reply python/tvm/driver/tvmc/autotuner.py:79]: I aggregated the discussion of remote tuning here and resolve other places to better track the states. Per @FrozenGene 's suggestions of using `--tracker-key` and `--rpc-device-key`, I think `rpc-device-key` is fine, but `tracker-key` is not a key so I would still prefer `rpc-tracker`. I think we can expect them to know the meaning of "tracker" and "device key", because if users need to run RPC runner, they must know how to launch RPC tracker and RPC client in advance.

In summary, here are the action items IMO:
1. Use `--rpc-tracker=hostname[:port]` and `--rpc-device-key=key-name`.
2. Have a checker to make sure both are specified or bot are not specified.
3. Explain how to launch RPC tracker and RPC client in the TVMC tutorial.

@leandron please share your thoughts on this.

Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:79]: I think it is good to have a more meaningful name, replacing the generic `--hostname / --port / --tracker-key` with something with more context like `--rpc-****`.

Just in the interest of making the parameters shorter and easier to remember, do you think `--rpc-key` would be acceptable, rather than `--rpc-device-key`? I find that two-word mnemonic are usually simpler to remember. I'm not sure it preserves the whole context you might want to give it.

Example of a command line:
```
tvmc tune --output log1.txt --target="llvm" --rpc-tracker=192.168.0.150:9191 --rpc-key=rasp4b resnet50-v2-7.onnx
```

Comment (FrozenGene) [reply python/tvm/driver/tvmc/autotuner.py:79]: `--rpc-key` is acceptable for me.

File: python/tvm/driver/tvmc/autotuner.py (hunk 3 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:91]: It refers to the compilation timeout. I've adjusted the help message to make it clear.

File: python/tvm/driver/tvmc/autotuner.py (hunk 5 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:134]: I suppose here you mean the empty blank line (134). Please let me know if I misunderstood.

File: python/tvm/driver/tvmc/autotuner.py (hunk 9 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:212]: I suppose here you mean the empty blank line (212). Please let me know if I misunderstood.

File: python/tvm/driver/tvmc/autotuner.py (hunk 8 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:187]: Makes sense. Removed.

File: python/tvm/driver/tvmc/autotuner.py (hunk 11 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:225]: Removed

File: python/tvm/driver/tvmc/autotuner.py (hunk 12 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:288]: Yep. Done

File: python/tvm/driver/tvmc/common.py (hunk 14 above)
Comment (leandron) [reply python/tvm/driver/tvmc/common.py:20]: Yep. I also added a note to migrate the compiler module to the common logger, in a separate PR.

File: python/tvm/driver/tvmc/autotuner.py (hunk 13 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:291]: Done

**Comment by leandron** [comment]
> We could add `--num-threads` to control how many cores to tune. This is important for remote devices like ARM

@FrozenGene that looks interesting. Just to clarify, is this not related to the parameter `n_parallel`, from `autotvm.RPCRunner`, already supported via `--parallel`?

> We could add `--cpu-mode` to control BIG core or LITTLE core to tune. Imagine remote devices like have A72 + A35. This feature maybe should require us to expose one interface in our AutoTVM to use API `runtime.config_threadpool`.

Yes, that looks useful. Can we add it in a separate PR?

File: python/tvm/driver/tvmc/autotuner.py (hunk 10 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:210]: looking into this

**Comment by FrozenGene** [comment]
> > We could add `--num-threads` to control how many cores to tune. This is important for remote devices like ARM
> 
> @FrozenGene that looks interesting. Just to clarify, is this not related to the parameter `n_parallel`, from `autotvm.RPCRunner`, already supported via `--parallel`?
> 

Yes, it is not related with `n_parallel`, which is to support we have multi devices to tune.

> > We could add `--cpu-mode` to control BIG core or LITTLE core to tune. Imagine remote devices like have A72 + A35. This feature maybe should require us to expose one interface in our AutoTVM to use API `runtime.config_threadpool`.
> 
> Yes, that looks useful. Can we add it in a separate PR, considering it requires some infra to be exposed to AutoTVM?

Fine to me.

**Comment by leandron** [comment]
> > @FrozenGene that looks interesting. Just to clarify, is this not related to the parameter `n_parallel`, from `autotvm.RPCRunner`, already supported via `--parallel`?
> 
> Yes, it is not related with `n_parallel`, which is to support we have multi devices to tune.

Understood. I'm happy to look into it. Can you expand a bit on how that would work, so that we can decide whether adding it here or on a future PR?

**Comment by FrozenGene** [comment]
> > > @FrozenGene that looks interesting. Just to clarify, is this not related to the parameter `n_parallel`, from `autotvm.RPCRunner`, already supported via `--parallel`?
> > 
> > 
> > Yes, it is not related with `n_parallel`, which is to support we have multi devices to tune.
> 
> Understood. I'm happy to look into it. Can you expand a bit on how that would work, so that we can decide whether adding it here or on a future PR?

You need to do the similar thing like we control BIG / LITTLE core via `runtime.config_threadpool`, this api could let us control how many cores and which cpu mode we like. So I am fine leave this in the future pr together with control BIG / LITTLE cpu mode.

File: python/tvm/driver/tvmc/autotuner.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+        "--early-stopping",
+        type=int,
+        help="minimum number of trials before early stopping",
+    )
+    parser.add_argument("--hostname", help="hostname or IP address of the host machine")
+
+    is_x86 = platform.machine() in ("i386", "AMD64", "x86_64")
+    parser.add_argument(
+        "--min-repeat-ms",
+        default=0 if is_x86 else 1000,
```
Comment (comaniac) [inline python/tvm/driver/tvmc/autotuner.py:52]: The value should be determined by the target platform instead of the host, so it should be based on the `target`. You probably need to post-process this value.

Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:52]: You're correct. I'm just wondering how can we have a mapping from "target string" to architecture, including the use cases for RPC tracking, in which supposedly the remote would need to provide the architecture somehow. Any ideas?

Suggestion/brainstorming here: In case we don't have clever way to collect an architecture for the target (that we can shortly explain in the help), I'd suggest we set default as `0` for `LocalRunner` and `1000` for `RPCRunner`, and keep it customisable via command line.

Comment (comaniac) [reply python/tvm/driver/tvmc/autotuner.py:52]: It's not that complicate tho. We can simply have `0 if target.keys[0] == "cpu" else 1000`. Since we have enabled clflush and it can make the measurement much more accurate, we don't need to set this value to 1000. However, it is only for x86, so we just need to check whether the first target key is for x86 (i.e. `cpu`).

Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:52]: Fair enough. I'll update accordingly.

File: python/tvm/driver/tvmc/autotuner.py (hunk 2 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py]: I pushed a new commit that implements `--rpc-tracker` and `--rpc-key`, as discussed here.

File: python/tvm/driver/tvmc/autotuner.py (hunk 10 above)
Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py:236]: Based on the existing tutorials that I had a look, it seems sensible just to completely remove `target_host` from here (and below, line 244), rather than expose it via some `--target-host TARGET_HOST`. What do you think?

Comment (comaniac) [reply python/tvm/driver/tvmc/autotuner.py:236]: Let's expose it via `--target-host` with `llvm` as its default value.

Comment (leandron) [reply python/tvm/driver/tvmc/autotuner.py]: Understood. Done now.

**Comment by leandron** [comment]
Re-triggering CI

**Comment by leandron** [comment]
This PR is passing CI again.

@comaniac @FrozenGene do you mind having another look?

**Review by comaniac — APPROVED** [review]

**Comment by comaniac** [comment]
Thanks @leandron @FrozenGene
